core/templates/dev/head/services/AssetsBackendApiService.ts

```
@@ -33,29 +33,47 @@ oppia.factory('AssetsBackendApiService', [
       $http, $q, AudioFileObjectFactory, CsrfTokenService,
       FileDownloadRequestObjectFactory, ImageFileObjectFactory,
       UrlInterpolationService, DEV_MODE) {
-    if (!DEV_MODE && !GLOBALS.GCS_RESOURCE_BUCKET_NAME) {
-      throw Error('GCS_RESOURCE_BUCKET_NAME is not set in prod.');
-    }
+    var GCS_RESOURCE_BUCKET_NAME = null;
+    var ASSET_TYPE_AUDIO = 'audio';
+    var ASSET_TYPE_IMAGE = 'image';
+    var GCS_PREFIX = null;
+    var AUDIO_DOWNLOAD_URL_TEMPLATE = null;
+    var IMAGE_DOWNLOAD_URL_TEMPLATE = null;
 
     // List of filenames that have been requested for but have
     // yet to return a response.
     var _audioFilesCurrentlyBeingRequested = [];
     var _imageFilesCurrentlyBeingRequested = [];
+    var AUDIO_UPLOAD_URL_TEMPLATE =
+      '/createhandler/audioupload/<exploration_id>';
 
-    var ASSET_TYPE_AUDIO = 'audio';
-    var ASSET_TYPE_IMAGE = 'image';
+    var _initialize = function() {
+      if (DEV_MODE) {
+        AUDIO_DOWNLOAD_URL_TEMPLATE = (
+          '/assetsdevhandler/<exploration_id>/assets/audio/<filename>');
+        IMAGE_DOWNLOAD_URL_TEMPLATE = (
+          '/assetsdevhandler/<exploration_id>/assets/image/<filename>');
+        return $q.resolve();
+      }
 
-    var GCS_PREFIX = ('https://storage.googleapis.com/' +
-      GLOBALS.GCS_RESOURCE_BUCKET_NAME + '/exploration');
-    var AUDIO_DOWNLOAD_URL_TEMPLATE = (
-      (DEV_MODE ? '/assetsdevhandler' : GCS_PREFIX) +
-      '/<exploration_id>/assets/audio/<filename>');
-    var IMAGE_DOWNLOAD_URL_TEMPLATE = (
-      (DEV_MODE ? '/assetsdevhandler' : GCS_PREFIX) +
-      '/<exploration_id>/assets/image/<filename>');
+      if (GCS_RESOURCE_BUCKET_NAME) {
```

If you are calling initialize() at the outset, wouldn't GCS_RESOURCE_BUCKET_NAME be null?

If you are not calling initialize() only at the outset, why call the method name "initialize"?

---

initialize is called where the assets backend service is needed.
the check is done for caching purposes. On prod, after sending the request, the response is being cached and a further call to initialize shouldn't do anything again since the service has already been initialized.

---

No, that is messy. Initialization of a service should be done exactly once per page.

If you're exposing this method publicly for others to call, then you would need to raise an exception if it's called more than once.